Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -0,0 +1,337 @@
+        case NotificationProfileDidFinishSyncing:
+            // Only reload top sites if there the cache is dirty since the finish syncing
+            // notification is fired everytime the user re-enters the app from the background.
+            self.profile.history.areTopSitesDirty(withLimit: ASPanelUX.topSitesCacheSize) >>== { dirty in
+                if dirty {
+                    self.refreshTopSites(ASPanelUX.topSitesCacheSize)
+                }
+            }
+        case NotificationFirefoxAccountChanged, NotificationPrivateDataClearedHistory, NotificationDynamicFontChanged:
+            self.refreshTopSites(ASPanelUX.topSitesCacheSize)
+        default:
+            // no need to do anything at all
+            print("Dont have this notification type")
+        }
+    }
+
+    private func refreshTopSites(frecencyLimit: Int) {
+        dispatch_async(dispatch_get_main_queue()) {
+            // Reload right away with whatever is in the cache, then check to see if the cache is invalid.
+            // If it's invalid, invalidate the cache and requery. This allows us to always show results
+            // immediately while also loading up-to-date results asynchronously if needed.
+            self.reloadTopSitesWithLimit(frecencyLimit) >>> {
+                self.profile.history.updateTopSitesCacheIfInvalidated() >>== { dirty in
+                    if dirty {
+                        self.reloadTopSitesWithLimit(frecencyLimit)
+                    }
+                }
+            }
+        }
+    }
+
+    private func reloadTopSitesWithLimit(limit: Int) -> Success {
+        return self.profile.history.getTopSitesWithLimit(limit).bindQueue(dispatch_get_main_queue()) { result in
+            if let data = result.successValue {
+                self.topSites = data.asArray().map { site in
+                    if let favURL = site.icon?.url {
+                        return TopSiteItem(urlTitle: site.tileURL.extractDomainName(), faviconURL: NSURL(string: favURL)!, siteURL: site.tileURL)
+                    }
+                    return TopSiteItem(urlTitle: site.tileURL.extractDomainName(), faviconURL: nil, siteURL: site.tileURL)
+                }
+                self.topSiteHandler.content = self.topSites
+                self.topSiteHandler.urlPressedHandler = { [unowned self] url in
+                    self.showSiteWithURL(url)
+                }
+                self.topSiteHandler.currentTraits = self.traitCollection
+                self.tableView.reloadData()
+            }
+            return succeed()
+        }
+    }
+
+    private func reloadRecentHistoryWithLimit(limit: Int) -> Success {
+        return self.profile.history.getSitesByLastVisit(limit).bindQueue(dispatch_get_main_queue()) { result in
+            if let data = result.successValue {
+                self.history = data.asArray()
+                self.tableView.reloadData()
+            }
+            return succeed()
+        }
+    }
```

Comment:
Looks like this was borrowed from the existing top sites. It's a bit confusing the way we interweave between background/main queue processing when updating sites. I did a minor rewrite of this that might make it more clear. It keeps things that can be done on the background in the background while grouping all the potentially main thread related things in the uponQueue block.

```swift

    private func reloadRecentHistory() {
        self.profile.history.getSitesByLastVisit(ASPanelUX.historySize).uponQueue(dispatch_get_main_queue()) { result in
            self.history = result.successValue?.asArray() ?? self.history
            self.tableView.reloadData()
        }
    }

    private func reloadTopSites() {
        invalidateTopSites().uponQueue(dispatch_get_main_queue()) { result in
            let sites = result.successValue ?? []
            self.topSites = sites
            self.topSiteHandler.content = self.topSites
            self.topSiteHandler.urlPressedHandler = { [unowned self] url in
                self.showSiteWithURL(url)
            }
            self.topSiteHandler.currentTraits = self.traitCollection
            self.tableView.reloadData()
        }
    }

    private func invalidateTopSites() -> Deferred<Maybe<[TopSiteItem]>> {
        let frecencyLimit = ASPanelUX.topSitesCacheSize
        return self.profile.history.areTopSitesDirty(withLimit: frecencyLimit) >>== { dirty in
            if dirty || self.topSites.isEmpty {
                return self.profile.history.getTopSitesWithLimit(frecencyLimit) >>== { topSites in
                    return deferMaybe(topSites.flatMap(self.siteToItem))
                }
            }
            return deferMaybe(self.topSites)
        }
    }

    private func siteToItem(site: Site?) -> TopSiteItem? {
        guard let site = site else {
            return nil
        }

        guard let faviconURL = site.icon?.url else {
            return TopSiteItem(urlTitle: site.tileURL.extractDomainName(), faviconURL: nil, siteURL: site.tileURL)
        }

        return TopSiteItem(urlTitle: site.tileURL.extractDomainName(), faviconURL: NSURL(string: faviconURL)!, siteURL: site.tileURL)
    }
```